expression/builtin_regexp_util.go

```diff
@@ -0,0 +1,251 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package expression
+
+import (
+	"regexp"
+	"unicode/utf8"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/util/chunk"
+)
+
+// Parameters may be const or ignored by the user, so different situations should be considered
+// We can handle parameters more easily with this struct.
+//
+// isNull field shows if user ignores this param in sql
+// for example:
+//
+//	select regexp_like("123", "123", "m"), here isNull field is false for the third parameter
+//	select regexp_like("123", "123"), here isNull field is true for the third parameter
+type regexpParam struct {
+	constStrVal   string
+	constIntVal   int64
+	defaultIntVal int64 // default value when isNull is true
+	isConst       bool
+	isNull        bool
+	col           *chunk.Column
+}
+
+func (re *regexpParam) getCol() *chunk.Column {
+	return re.col
+}
+
+func (re *regexpParam) getStringVal(id int) string {
+	if re.isNull {
+		return ""
+	}
+
+	if re.isConst {
+		return re.constStrVal
+	}
+
+	return re.col.GetString(id)
+}
+
+func (re *regexpParam) getIntVal(id int) int64 {
+	if re.isNull {
+		return re.defaultIntVal
+	}
+
+	if re.isConst {
+		return re.constIntVal
+	}
+
+	return re.col.GetInt64(id)
+}
+
+func buildStringParam(bf *baseBuiltinFunc, id int, input *chunk.Chunk, isNull bool) (*regexpParam, error) {
+	var pa regexpParam
+	var err error
+
+	pa.isNull = isNull
+	if pa.isNull {
+		return &pa, nil
+	}
+
+	pa.col, err = bf.bufAllocator.get()
+	if err != nil {
+		return nil, err
+	}
+
+	// Get values from input
+	if err := bf.args[id].VecEvalString(bf.ctx, input, pa.col); err != nil {
+		return nil, err
+	}
+
+	// Check if this is a const value
+	pa.isConst = bf.args[id].ConstItem(bf.ctx.GetSessionVars().StmtCtx)
+	if pa.isConst {
+		// Initialize the const
+		pa.constStrVal, err = getColumnConstValString(pa.col, input.NumRows())
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &pa, nil
+}
+
+func buildIntParam(bf *baseBuiltinFunc, id int, input *chunk.Chunk, isNull bool, defaultIntVal int64) (*regexpParam, error) {
+	var pa regexpParam
+	var err error
+
+	pa.isNull = isNull
+	if pa.isNull {
+		pa.defaultIntVal = defaultIntVal
+		return &pa, nil
+	}
+
+	pa.col, err = bf.bufAllocator.get()
+	if err != nil {
+		return nil, err
+	}
+
+	// Get values from input
+	if err := bf.args[id].VecEvalInt(bf.ctx, input, pa.col); err != nil {
+		return nil, err
+	}
+
+	// Check if this is a const value
+	pa.isConst = bf.args[id].ConstItem(bf.ctx.GetSessionVars().StmtCtx)
+	if pa.isConst {
+		// Initialize the const
+		pa.constIntVal, err = getColumnConstValInt(pa.col, input.NumRows())
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &pa, nil
+}
+
+func getColumnConstValString(col *chunk.Column, n int) (string, error) {
+	// Precondition: col is generated from a constant expression
+	for i := 0; i < n; i++ {
+		if col.IsNull(i) {
+			continue
```

Comment:
It looks meaningless since if col is a constant column, and if one of the element is null, then all the element of the column should be null.